### DISCUSSION TITLE
Lo-res color oddness

### DISCUSSION THREAD

**Discussion by epaell** [markadev/AppleII-VGA#11] [Q&A] [answered]
I've been trying to understand why my colours seem a bit off when running in lores mode. I've checked the output circuitry and I'm fairly certain that all the components are in the correct locations and that the wiring is correct (taking the right outputs from the pico pi going via the resistors and into the VGA output connector).

When I run a simple basic program to draw bars with each of the 16 possible colours and take the composite output I get something like this:
![composite](https://github.com/markadev/AppleII-VGA/assets/7944844/633e869e-523f-4185-972f-18f5fb033e07)

When I then connect to the VGA output I get something like this:
![vga](https://github.com/markadev/AppleII-VGA/assets/7944844/35ce8d23-6703-4b61-99e4-cdb73d0a37d1)

The VGA output is also similar if I don't have it plugged into the Apple II+ and instead write a test program for the pico pi to make the same pattern.

When I generate the original AppleII-VGA test pattern I get something that looks quite similar to what is shown in the documentation e.g.

![test_pattern](https://github.com/markadev/AppleII-VGA/assets/7944844/27f13f7b-dfbf-411c-b538-5eae03af116a)

I've tried to hook up via a VGA-to-HDMI cable to our living room TV and another monitor and the result is similar.

It's not a major issue but just really curious as to what could be doing this (particularly for colour 1 which should be magenta but ends up looking more like purple/violet).

**Comment by epaell** [comment]
I ran one more test where I modified the original test pattern in the pico pi to generate steps in the individual primary colours (red, green and blue) in the three central squares - just to be absolutely certain that they generate the expected colours and that they step up in brightness as one would expect. This seems to be as one would expect:
![test_rgb](https://github.com/markadev/AppleII-VGA/assets/7944844/00649e16-368d-43cb-8f5d-d968ee6376e2)

Something that came to mind is whether the proportion of the colours is correct when they are combined. Could slightly wrong resistor values have a great effect on the resultant colour? I think I used 470 ohm instead of 500 ohm and the other resistors will have values which will be out by their level of tolerance. Though, I guess I wouldn't have expected this to have a significant effect.

**Reply by markadev** [reply]
The resistor value could be contributing a little but I think it's just that the palette values defined in `render_lores.c` are not very accurate to the original.  The "magenta" in particular I've seen described as "red" in some places or "magenta" in others and I think I erred on the side of magenta/purple too much.

For example, according to [this old post](http://mrob.com/pub/xapple2/colors.html), the RGB values for magenta (labeled as "red" 😂) should theoretically be 227,30,96.  And the Apple IIGS (a proper RGB system!) calls it "deep red" and uses 221, 0, 51.

```
..Original Apple ][ values...  ..........IIGS values..........
                   -R- -G- -B-              12-bit  -R- -G- -B-
 black    COLOR=0    0   0   0  Black       $0000     0   0   0
 red      COLOR=1  227  30  96  Deep Red    $0D03   221   0  51
 dk blue  COLOR=2   96  78 189  Dark Blue   $0009     0   0 153
 purple   COLOR=3  255  68 253  Purple      $0D2D   221  34 221
 dk green COLOR=4    0 163  96  Dark Green  $0072     0 119  34
 gray     COLOR=5  156 156 156  Dark Gray   $0555    85  85  85
 med blue COLOR=6   20 207 253  Medium Blue $022F    34  34 255
 lt blue  COLOR=7  208 195 255  Light Blue  $06AF   102 170 255
 brown    COLOR=8   96 114   3  Brown       $0850   136  85   0
 orange   COLOR=9  255 106  60  Orange      $0F60   255 102   0
 grey     COLOR=10 156 156 156  Light Gray  $0AAA   170 170 170
 pink     COLOR=11 255 160 208  Pink        $0F98   255 153 136
 lt green COLOR=12  20 245  60  Light Green $01D0    17 221   0
 yellow   COLOR=13 208 221 141  Yellow      $0FF0   255 255   0
 aqua     COLOR=14 114 255 208  Aquamarine  $04F9    68 255 153
 white    COLOR=15 255 255 255  White       $0FFF   255 255 255
 ```

My low-quality Dell LCD probably didn't do me any favors when I was putting the initial palette together either since the colors shift drastically based on the viewing angle.

**Reply by epaell** [reply]
I wonder if "Never The Same Color" (NTSC) is also coming into play as well, at least when comparing against the composite output :-) OK, that's good to know - I thought maybe it was just me and paranoia was starting to settle in (my super-cheap mini "HD" monitor also suffers greatly from viewing angle issues - I had a lot of trouble taking pictures of it). @Paco1979 has fiddled with the colours the other branch so I might try those on for size and see how they compare.

**Reply by markadev** [reply, marked as answer]
I double-checked my card's composite output and it does pretty much match yours, where the "magenta" is really more red, but my "orange" is uglier than yours and my "yellow" is nicer.  NTSC indeed 😂.

I tried the RGB values from the Apple IIGS palette and the VGA out looks much nicer so I'm going to commit the new palette to the code.

Composite:
![PXL_20230610_172146696](https://github.com/markadev/AppleII-VGA/assets/17986295/19147cfd-d28d-4ea5-baaa-6da035bb8dd1)

IIGS palette:
![PXL_20230610_173321355](https://github.com/markadev/AppleII-VGA/assets/17986295/58123f2c-256a-4fc5-8e35-f8c4f069e7ee)

**Reply by epaell** [reply]
That's looking great!
